### ISSUE TITLE
Contract negotiation loop

### ISSUE THREAD

**Issue by hlange** [eclipse-dataspaceconnector/DataSpaceConnector#538]
Hi,
I have a problem with the sample 4.0. I fixed the issue with keyName like mentioned in the discussions. But this generated new problems because after everything gone through and the file was moved from one directory to a different, the CPU load of one core gone up to 100%.

Because I thought maybe this could be because it was not the newest version I clone the current version. This generated a complete new issue because the negotiation was running in a loop.

Consumer:
![grafik](https://user-images.githubusercontent.com/12150189/150515320-793bd0f6-466a-4e2d-a458-bc5e27c73ddc.png)

Provider:
![grafik](https://user-images.githubusercontent.com/12150189/150515377-96bf420d-b206-4139-b55c-4e0e1acafa5b.png)

By running the following command:
`curl -X POST -H "Content-Type: application/json" -d @samples/04.0-file-transfer/contractoffer.json "http://localhost:9191/api/negotiation?connectorAddress=http://localhost:8181/api/ids/multipart"`

And I added keyName to ConsumerApiController.java

Best regards,
Henrik

**Comment by ndr-brt** [comment]
Thanks for finding out, here I see two problems:
* the sample 04 was renamed to 04.0 but its README was not updated accordingly
* on the consumer side ContractNegotiation `REQUESTING` state has two meanings:
  * ContractNegotiation created and waiting to be sent to the provider
  * ContractNegotiation sent to the provider and waiting for response

This causes the infinite loop, because every `REQUESTING` negotiation is sent to the provider, but it should not

**Comment by ndr-brt** [comment]
I noticed strange behavior into the contract negotiation repositories (stores).
In fact, both of the "production" (I mean "production code", not "production ready") implementations contain this logic:
```
if (ContractNegotiationStates.UNSAVED.code() == negotiation.getState()) {
            if (ContractNegotiation.Type.CONSUMER.equals(negotiation.getType())) {
                negotiation.transitionRequesting();
            } else {
                negotiation.transitionRequested();
            }
        }
```

this is dangerous because this kind of domain logic should stay in the domain classes, in this case in `ContractNegotiation` probably. Otherwise every time someone implement a `ContractNegotiationStore` should report that logic in their implementation, and that would bring to duplication.

However my proposal is to define an additional state between UNSAVED and REQUESTING called INITIALIZED. The negotiations on consumer state in that state will be sent to provider and moved to PROCESSING. That would resolve this bug

**Comment by mspiekermann** [comment]
closed by #545
